### A/B Hypothesis Testing
    Accept or reject the following Null Hypotheses: 
        There are no risk differences across provinces 
        There are no risk differences between zip codes 
        There are no significant margin (profit) difference between zip codes 
        There are not significant risk difference between Women and Men


### A/B Hypothesis Testing: Risk Differences Across Provinces
### Null Hypothesis (H0):
* There are no risk differences across provinces.
### Alternative Hypothesis (H1):
* There are risk differences across provinces.

### Selecting a Metric
* To measure risk, we can use the loss ratio. This is calculated as:

* Loss Ratio = Total Claims Paid / Total Premiums Earned
* A higher loss ratio indicates higher risk.

### Data Segmentation
* Group A: Policyholders from a randomly selected subset of provinces.
* Group B: Policyholders from the remaining provinces.
### Statistical Testing
* Chi-squared test: Compare the distribution of loss ratios between Group A and Group B.

In [1]:
import os 
os.chdir("../")

In [2]:
import pandas as pd
df = pd.read_csv("data/dataset.csv")

C:\Users\windows 10\AppData\Local\Temp\ipykernel_12144\685072172.py:2: DtypeWarning: Columns (33,38) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("data/dataset.csv")


##### Maximum loss ratio:
* 2120643.85964912
##### Minimum loss ratio:
* -0.12709600000000001
#### Mean loss ratio:
* 128.8991 (i use this value as threshold)

In [3]:
from scripts.preprocessing import replace_missing_with_mean, replace_missing_with_mode

df = df.drop(columns=['NumberOfVehiclesInFleet','CrossBorder' ])
numerical_cols = df.select_dtypes(include=['int64', 'float64'])
categorical_cols = df.select_dtypes(include='object')

df = replace_missing_with_mode(df,categorical_cols)
df = replace_missing_with_mean(df,numerical_cols.columns)


In [4]:
from scripts.Hypothesis_Testing import AB_haypothe
chi2, p_value, dof, expected = AB_haypothe(df,"Province")
print('The chi2 value',chi2)
print('The p vlaue', p_value)
print('The dof value', dof)
print('The expected value',expected)
if p_value < 0.05:
    print("Reject null hypothesis: There are significant risk differences across provinces.")
else:
    print("Fail to reject null hypothesis: There are no significant risk differences across provinces.")

The chi2 value 94.82137297811462
The p vlaue 4.8644771983937436e-17
The dof value 8
The expected value [[3.03106416e+04 2.53584109e+01]
 [8.09222990e+03 6.77010053e+00]
 [3.93535761e+05 3.29238875e+02]
 [1.69639077e+05 1.41923008e+02]
 [2.48152391e+04 2.07608614e+01]
 [5.26739321e+04 4.40679293e+01]
 [1.43167224e+05 1.19776194e+02]
 [6.37466684e+03 5.33315735e+00]
 [1.70653229e+05 1.42771464e+02]]
Reject null hypothesis: There are significant risk differences across provinces.


### Result
#### The chi2 value:
* 94.82137297811462
#### The p vlaue:
* 4.8644771983937436e-17
#### The dof value:
* 8
#### The expected value:
* [[3.03106416e+04 2.53584109e+01]
* [8.09222990e+03 6.77010053e+00]
* [3.93535761e+05 3.29238875e+02]
* [1.69639077e+05 1.41923008e+02]
* [2.48152391e+04 2.07608614e+01]
* [5.26739321e+04 4.40679293e+01]
* [1.43167224e+05 1.19776194e+02]
* [6.37466684e+03 5.33315735e+00]
* [1.70653229e+05 1.42771464e+02]]
### Reject null hypothesis:
* There are significant risk differences across provinces.

### Understanding the Output:

#####chi2 value (94.82137297811462):
* This is the calculated chi-squared test statistic. A higher value indicates a greater difference between the observed and expected frequencies.
#### p-value (4.8644771983937436e-17):
* This is the probability of observing a test statistic as extreme or more extreme as the calculated chi-squared value, assuming the null hypothesis is true. A very low p-value suggests that the observed differences are unlikely to be due to chance.
#### dof value (8):
* This is the degrees of freedom for the chi-squared test, which depends on the size of your contingency table.
#### expected value matrix:
* This matrix contains the expected frequencies under the assumption of independence between the provinces and the high-risk status.
### Conclusion:

* Given the low p-value (4.8644771983937436e-17), you can reject the null hypothesis and conclude that there are significant risk differences across provinces. This means that the observed differences in loss ratios between the provinces are unlikely to be due to chance.

### There are no risk differences between zip codes 


### A/B Hypothesis Testing : There are no risk differences between zip codes 

#### Null Hypothesis (H0):

* There are no risk differences between zip codes.
#### Alternative Hypothesis (H1):

* There are risk differences between zip codes.


In [5]:

chi2_zip, p_value_zip, dof_zip, expected_zip = AB_haypothe(df,"PostalCode")
print('The chi2_zip value',chi2_zip)
print('The p_value_zip', p_value_zip)
print('The dof_zip value', dof_zip)
print('The expected_zip value',expected_zip)

# Interpret results for zip codes
if p_value_zip < 0.05:
    print("Reject null hypothesis: There are significant risk differences between zip codes.")
else:
    print("Fail to reject null hypothesis: There are no significant risk differences between zip codes.")

The chi2_zip value 1256.157467912241
The p_value_zip 3.241824585855775e-15
The dof_zip value 887
The expected_zip value [[5.33653536e+03 4.46463847e+00]
 [1.48675615e+03 1.24384610e+00]
 [7.69356343e+01 6.43656922e-02]
 ...
 [9.99164082e+01 8.35918080e-02]
 [1.41381718e+03 1.18282408e+00]
 [2.19816098e+02 1.83901978e-01]]
Reject null hypothesis: There are significant risk differences between zip codes.


### value 
#### The chi2_zip value:
* 1256.157467912241
#### The p_value_zip:
* 3.241824585855775e-15
#### The dof_zip value:
* 887
#### The expected_zip value:
* [[5.33653536e+03 4.46463847e+00]
* [1.48675615e+03 1.24384610e+00]
* [7.69356343e+01 6.43656922e-02]
* ...
* [9.99164082e+01 8.35918080e-02]
* [1.41381718e+03 1.18282408e+00]
* [2.19816098e+02 1.83901978e-01]]
### Reject null hypothesis:
* There are significant risk differences between zip codes.

### Understanding the Output:

#### chi2_zip value (1256.157467912241):
* This is the calculated chi-squared test statistic. A higher value indicates a greater difference between the observed and expected frequencies.
#### p_value_zip (3.241824585855775e-15):
* This is the p-value associated with the test statistic. A very low p-value suggests that the observed differences are unlikely to be due to chance.
#### dof_zip value (887):
* This is the degrees of freedom for the chi-squared test, which depends on the size of your contingency table.
#### expected_zip value:
* This matrix contains the expected frequencies under the assumption of independence between zip codes and the high-risk status.
####Conclusion:

* Given the low p-value (3.241824585855775e-15), you can reject the null hypothesis and conclude that there are significant risk differences between zip codes. This means that the observed differences in loss ratios between zip codes are unlikely to be due to chance.

### Hypothesis Testing: There are no significant margin (profit) difference between zip codes 
#### Null Hypothesis (H0):

* There are no significant profit margin differences between zip codes.
#### Alternative Hypothesis (H1):

* There are significant profit margin differences between zip codes.
##### Calculating Profit Margin
* To calculate profit margin, you'll need to have information about the premiums earned and expenses incurred for each policy.

* Profit Margin = (Total Premiums Earned - Total Expenses) / Total Premiums 


In [6]:
from scripts.Hypothesis_Testing import calculate_profit_margin_and_anova


f_statistic, p_value = calculate_profit_margin_and_anova(df)
print('The f stastistic value ',f_statistic)
print('The p value ', p_value)

# Interpret results
if p_value < 0.05:
    print("Reject null hypothesis: There are significant profit margin differences between zip codes.")
else:
    print("Fail to reject null hypothesis: There are no significant profit margin differences between zip codes.")

The f stastistic value  0.8252594843896949
The p value  0.9999431392284237
Fail to reject null hypothesis: There are no significant profit margin differences between zip codes.


### Result 
#### The f stastistic value:
*  0.8252594843896949
#### The p value:
*  0.9999431392284237
### Fail to reject null hypothesis:
* There are no significant profit margin differences between zip codes.

### Interpreting the Results:

#### F-statistic (0.8252594843896949):
* This is the calculated F-statistic, which measures the variation between group means relative to the variation within groups. A higher F-statistic indicates greater differences between groups.
#### p-value (0.9999431392284237):
* This is the probability of observing an F-statistic as extreme or more extreme as the calculated value, assuming the null hypothesis (no difference between groups) is true. A high p-value suggests that the observed differences are likely due to chance.

### Conclusion:

* Given the high p-value (0.9999431392284237), you fail to reject the null hypothesis. This means that there is not enough statistical evidence to conclude that there are significant profit margin differences between zip codes.

### Hypothesis Testing: There are not significant risk difference between Women and Men
#### Null Hypothesis (H0):

* There are no significant risk differences between women and men.
#### Alternative Hypothesis (H1):

* There are significant risk differences between women and men.
#### Data Preparation
* Calculate Loss Ratio: As we did previously, calculate the loss ratio for each policy.
* Group by Gender: Group the data by gender.
#### Statistical Testing
* T-Test: Use an independent t-test to compare the mean loss ratios between women and men.


In [7]:
from scripts.Hypothesis_Testing import calculate_loss_ratio_and_t_test_for_gender

t_stat, p_value = calculate_loss_ratio_and_t_test_for_gender(df, "Gender", "LossRatio")

print('The t stat value ', t_stat)
print('The p value ', p_value)
# Interpret results
if p_value < 0.05:
    print("Reject null hypothesis: There are significant risk differences between women and men.")
else:
    print("Fail to reject null hypothesis: There are no significant risk differences between women and men.")

The t stat value  1.275317702578255
The p value  0.20220279120352017
Fail to reject null hypothesis: There are no significant risk differences between women and men.


### Result

#### The t stat value:
*  1.275317702578255
#### The p value:
*  0.20220279120352017
#### Fail to reject null hypothesis:
* There are no significant risk differences between women and men.

### Understanding the Output:

#### t stat value (1.275317702578255):
* This is the calculated t-statistic, which measures the difference between the means of the two groups (women and men) relative to the standard error.
#### p value (0.20220279120352017):
* This is the probability of observing a test statistic as extreme or more extreme as the calculated t-statistic, assuming the null hypothesis is true.

### Conclusion:

* Given the p-value of 0.20220279120352017, which is greater than the typical alpha level of 0.05, you fail to reject the null hypothesis. This means that there is not enough statistical evidence to conclude that there are significant risk differences between women and men.

